### PR TITLE
added TextChunkingConfig to FileSplitterConfig

### PR DESCRIPTION
**Pull request by QuantumPlumber**
## Description

Added abatch support for document contextual pre-processing.

## Type of change

(Please delete options that are not relevant)

- [x] ⚡ New feature (non-breaking change which adds functionality)

## Checklist:

- [ x] My code follows the style guidelines(PEP 8) of MindsDB.
- [ x] I have appropriately commented on my code, especially in complex areas.
- [ x] Necessary documentation updates are either made or tracked in issues.
- [ x] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Comment by mindsdbadmin** [comment]
<br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/mindsdb/mindsdb/blob/main/assets/contributions-agreement/individual-contributor.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   <sub>You can retrigger this bot by commenting **recheck** in this Pull Request. </sub><sub>Posted by the **CLA Assistant Lite bot**.</sub>

**Review by dusvyat — APPROVED** [review]
lgtm

File: mindsdb/integrations/utilities/rag/splitters/file_splitter.py [hunk 1]
```diff
@@ -2,80 +2,109 @@
...
+from mindsdb.interfaces.knowledge_base.preprocessing.models import TextChunkingConfig
 
 from mindsdb.utilities import log
 
 DEFAULT_CHUNK_SIZE = 1000
 DEFAULT_CHUNK_OVERLAP = 50
 DEFAULT_MARKDOWN_HEADERS_TO_SPLIT_ON = [
-    ('#', 'Header 1'),
-    ('##', 'Header 2'),
-    ('###', 'Header 3')
```
Comment (dusvyat) [inline mindsdb/integrations/utilities/rag/splitters/file_splitter.py:14]: nit: would rather put formatting changes in a separate PR to make it clearer what functionality is being change

File: mindsdb/interfaces/knowledge_base/preprocessing/document_preprocessor.py [hunk 2]
```diff
@@ -119,23 +142,36 @@ def __init__(self, config: ContextualConfig):
...
 
-    def _generate_context(self, chunk_content: str, full_document: str) -> str:
+    def _generate_context(
+        self, chunk_contents: list[str], full_documents: list[str]
+    ) -> list[str]:
         """Generate contextual description for a chunk using LLM"""
-        prompt = self.context_template.replace("{{WHOLE_DOCUMENT}}", full_document)
-        prompt = prompt.replace("{{CHUNK_CONTENT}}", chunk_content)
-        response = self.llm(prompt)
+        prompts = [
```
Comment (dusvyat) [inline mindsdb/interfaces/knowledge_base/preprocessing/document_preprocessor.py:165]: AttributeError: 'list' object has no attribute 'replace' 

You can run [tests/unit/test_document_preprocessor.py](https://github.com/mindsdb/mindsdb/blob/ed7cdb2fa3ff715a1e8fa3fab649418e937929a7/tests/unit/test_document_preprocessor.py) to validate that changes haven't broken anything

**Review by dusvyat — CHANGES_REQUESTED** [review]
see comment on error, otherwise I think looks good

File: mindsdb/integrations/utilities/rag/splitters/file_splitter.py [hunk 3]
```diff
@@ -2,80 +2,109 @@
...
 class FileSplitterConfig:
-    '''Represents configuration needed to split a file into chunks for retrieval.'''
+    """Represents configuration needed to split a file into chunks for retrieval."""
+
     # Target chunk size in characters. Not all splitters will adhere exactly to this (it's more of a guideline)
     chunk_size: int = DEFAULT_CHUNK_SIZE
     # How many characters each chunk should overlap. Not all splitters will adhere exactly to this (it's more of a guideline)
     chunk_overlap: int = DEFAULT_CHUNK_OVERLAP
+    # Chunking parameters are passed as a TextChunkingConfig
+    text_chunking_config: TextChunkingConfig = TextChunkingConfig(
```
Comment (dusvyat) [inline mindsdb/integrations/utilities/rag/splitters/file_splitter.py:40]: if we add thus then having chunk_size and chunk_ovlerlap in addition becomes redundant

File: mindsdb/integrations/utilities/rag/splitters/file_splitter.py [hunk 4]
```diff
@@ -2,80 +2,109 @@
...
+        chunk_size=DEFAULT_CHUNK_SIZE, chunk_overlap=DEFAULT_CHUNK_OVERLAP
+    )
     # Default recursive splitter to use for text files, or unsupported files
-    recursive_splitter: RecursiveCharacterTextSplitter = RecursiveCharacterTextSplitter(chunk_size=chunk_size, chunk_overlap=chunk_overlap)
+    recursive_splitter: RecursiveCharacterTextSplitter = RecursiveCharacterTextSplitter(
```
Comment (dusvyat) [inline mindsdb/integrations/utilities/rag/splitters/file_splitter.py:44]: potentially, text_chunking_config is not full initialised, may be better to move to a  `__post_init__`

 ```
 .
 .
recursive_splitter: RecursiveCharacterTextSplitter = None

    def __post_init__(self):
        if self.recursive_splitter is None:
            self.recursive_splitter = RecursiveCharacterTextSplitter(
                chunk_size=self.text_chunking_config.chunk_size,
                chunk_overlap=self.text_chunking_config.chunk_overlap,
                length_function=self.text_chunking_config.length_function,
                separators=self.text_chunking_config.separators,
            )
```

File: mindsdb/interfaces/knowledge_base/preprocessing/document_preprocessor.py [hunk 5]
```diff
@@ -119,23 +142,36 @@ def __init__(self, config: ContextualConfig):
...
+            self.context_template.replace("{{WHOLE_DOCUMENT}}", full_document)
+            for full_document in full_documents
+        ]
+        prompts = [
+            prompts.replace("{{CHUNK_CONTENT}}", chunk_content)
```
Comment (dusvyat) [inline mindsdb/interfaces/knowledge_base/preprocessing/document_preprocessor.py:170]: `AttributeError: 'list' object has no attribute 'replace'`

You can run [tests/unit/test_document_preprocessor.py](https://github.com/mindsdb/mindsdb/blob/ed7cdb2fa3ff715a1e8fa3fab649418e937929a7/tests/unit/test_document_preprocessor.py) to validate that changes haven't broken anything

**Comment by QuantumPlumber** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by QuantumPlumber** [comment]
Closing to re-open

**Pull request closed without merging**